### PR TITLE
A client for talking to the Java Spalloc Server

### PR DESCRIPTION
**Pull request by dkfellows**
This code was originally in SpiNNakerManchester/SpiNNFrontEndCommon#842, but the adding of the proxying capability made it clear that it actually belongs here instead.

See also SpiNNakerManchester/JavaSpiNNaker#444

### CODE REVIEW

**Comment by dkfellows** [comment]
This PR also changes the name and docstring of `DatabaseConfirmation` to make it clearer that that message is not routed via SpiNNaker at all.

File: spinnman/spalloc/abstract_classes.py [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+from .enums import SpallocState
+
+_ONE_SHORT = struct.Struct("<H")
+_TWO_SHORTS = struct.Struct("<2H")
+_TWO_SKIP: bytes = b'\0\0'
+_NUM_UPDATE_TAG_TRIES = 3
+_UPDATE_TAG_TIMEOUT = 1.0
+
+
+class AbstractSpallocClient(object, metaclass=AbstractBase):
```
Comment (rowleya) [inline spinnman/spalloc/abstract_classes.py:53]: I would suggest that this and the other classes in this file are broken out into files of their own.  This makes them easier to find in the future...

File: spinnman/spalloc/enums.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from enum import IntEnum
+
+
+class SpallocState(IntEnum):
```
Comment (rowleya) [inline spinnman/spalloc/enums.py:18]: Similarly, these enums would be better split out, and this is more consistent with the style in the rest of the code.

File: spinnman/spalloc/spalloc.py [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+_close_req = struct.Struct("<III")
+_open_listen_req = struct.Struct("<II")
+# Open and close share the response structure
+_open_close_res = struct.Struct("<III")
+_open_listen_res = struct.Struct("<IIB4I")
+_msg = struct.Struct("<II")
+_msg_to = struct.Struct("<IIII")
+
+
+class SpallocClient(AbstractContextManager, AbstractSpallocClient):
```
Comment (rowleya) [inline spinnman/spalloc/spalloc.py:57]: Maybe rename this spalloc_client.py as well.  Although it has other classes in the file, those are private to this module and so don't need to be broken out, but the main thing in this file is SpallocClient.

File: spinnman/spalloc/abstract_classes.py (hunk 1 above)
Comment (dkfellows) [reply spinnman/spalloc/abstract_classes.py]: Done

File: spinnman/spalloc/enums.py (hunk 2 above)
Comment (dkfellows) [reply spinnman/spalloc/enums.py]: Done

File: spinnman/spalloc/spalloc.py (hunk 3 above)
Comment (dkfellows) [reply spinnman/spalloc/spalloc.py]: Done

**Review by rowleya — APPROVED** [review]
This is fine by me now, but I have also worked on it, so if anyone else wants to review, please go ahead!

**Review by andrewgait — APPROVED** [review]
Looks fine to me.

**Pull request merged**